packages/runtime/container-runtime/src/test/containerRuntime.spec.ts

```diff
@@ -1465,5 +1513,172 @@ describe("Runtime", () => {
+						// Submit an op and yield for it to be flushed from outbox to pending state manager.
+						containerRuntime.submitDataStoreOp("fakeId", "fakeContents");
+						await yieldEventLoop();
+						return boundFn(...args);
+					};
+				};
+				containerRuntime.summarize = patch(containerRuntime.summarize);
+
+				const summarizeResult = await containerRuntime.submitSummary({
+					summaryLogger: createChildLogger(),
+					cancellationToken: neverCancelledSummaryToken,
+				});
+				assert(summarizeResult.stage === "base", "Summary did not fail");
+				assert.strictEqual(
+					summarizeResult.error.message,
+					"PendingOpsWhileSummarizing",
+					"Summary did not fail with the right error",
+				);
+				assert.strictEqual(
+					summarizeResult.error.beforeGenerate,
+					false,
+					"It should have failed after generating summary",
+				);
+			});
+
+			it("summary passes if pending ops are processed during pending op processing timeout", async () => {
+				// Create a container runtime type where the submit method is public. This makes it easier to test
+				// submission and processing of ops. The other option is to send data store or alias ops whose
+				// processing requires creation of data store context and runtime as well.
+				type ContainerRuntimeWithSubmit = Omit<ContainerRuntime, "submit"> & {
+					submit(
+						containerRuntimeMessage: ContainerRuntimeMessage,
+						localOpMetadata: unknown,
+						metadata: Record<string, unknown> | undefined,
+					): void;
+				};
+				const containerRuntimeWithSubmit =
+					containerRuntime as unknown as ContainerRuntimeWithSubmit;
+				// Submit a rejoin op and yield for it to be flushed from outbox to pending state manager.
+				containerRuntimeWithSubmit.submit(
+					{
+						type: ContainerMessageType.Rejoin,
+						contents: "something",
+					},
+					undefined,
+					undefined,
+				);
+				await yieldEventLoop();
+
+				// Create a mock logger to validate that pending ops event is generated with correct params.
+				const mockLogger = new MockLogger();
+				const summaryLogger = createChildLogger({ logger: mockLogger });
+				const summarizeResultP = containerRuntime.submitSummary({
+					summaryLogger,
+					cancellationToken: neverCancelledSummaryToken,
+				});
+
+				// Advance the clock by 1 ms less than the time waited for pending ops to be processed. This will allow
+				// summarization to proceed far enough to wait for pending ops.
+				clock.tick(defaultPendingOpsWaitTimeoutMs - 1);
+				// Process the rejoin op so that there are no pending ops.
+				containerRuntime.process(
+					{
+						type: "op",
+						clientId: "fakeClientId",
+						sequenceNumber: 0,
+						contents: {
+							type: ContainerMessageType.Rejoin,
+							contents: "something",
+						},
+					} as any as ISequencedDocumentMessage,
+					true /* local */,
+				);
+				// Advance the clock by the remaining time so that pending ops wait is completed.
+				clock.tick(1);
```

Comment:
Is this necessary?

Reply:
yeah, because the previoud tick in line 1654 did 1 ms less than the timeout.

Reply:
But why do we need to hit the timeout?  Just to prove that it won't actually fire, since the last (only) pending op was processed?

Reply:
Not sure if I missing soemthing but if the timeout doesn't fire, the submitSummary call won't complete right? These tests use fake timers so we have to manually advance them otherwise submitSummary will be stuck waiting for pending ops.

Reply:
The Promise where the Timeout is also has this:

```ts
this.once("saved", () => {
					clearTimeout(timeoutId);
					resolve();
				});
```

Shouldn't processing that pending op trigger "saved" event?  I didn't track through the code to verify that, but that's what I thought that event was about.  In that case the promise will resolve even if the timeout didn't yet fire.